src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java

```diff
@@ -136,7 +147,46 @@ public List<LinkedFile> findAssociatedNotLinkedFiles(BibEntry entry) throws IOEx
                 linkedFiles.add(linkedFile);
             }
         }
-
         return linkedFiles;
     }
+
+    public RelinkedResults relinkingFiles(List<LinkedFile> listlinked) {
+        List<LinkedFile> changedFiles = new ArrayList<>();
+        List<IOException> exceptions = new ArrayList<>();
+
+        for (LinkedFile file : listlinked) {
+            Path path = Path.of(file.getLink());
+            if (!Files.exists(path)) {
+                Path filePath = Path.of(file.getLink());
+                String directoryPath = filePath.getParent().getParent().toString();
+                File directory = new File(directoryPath);
+
+                String fileNameString = filePath.getFileName().toString();
+
+                List<String> fileLocations = new ArrayList<>();
+
+                searchFileInDirectoryAndSubdirectories(directory, fileNameString, fileLocations);
+                if (!fileLocations.isEmpty()) {
+                    file.setLink(fileLocations.get(0));
+                    changedFiles.add(file);
+                } else {
+                    exceptions.add(new IOException("couldn't find file: " + file.getLink()));
+                }
+            }
+        }
+        return new RelinkedResults(changedFiles, exceptions);
+    }
+
+    public static void searchFileInDirectoryAndSubdirectories(File directory, String targetFileName, List<String> fileLocations) {
```

Comment:
you should use Path from nio here as well, see https://mkyong.com/java/java-files-walk-examples/ for an example